x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/VectorSimilarityFunctionsPushdown.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules.physical.local;
+
+import org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper;
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.FieldAttribute;
+import org.elasticsearch.xpack.esql.core.expression.FieldFunctionAttribute;
+import org.elasticsearch.xpack.esql.core.expression.Literal;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.vector.VectorSimilarityFunction;
+import org.elasticsearch.xpack.esql.optimizer.PhysicalOptimizerRules;
+import org.elasticsearch.xpack.esql.plan.physical.EvalExec;
+import org.elasticsearch.xpack.esql.plan.physical.FieldExtractExec;
+import org.elasticsearch.xpack.esql.plan.physical.PhysicalPlan;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+public class VectorSimilarityFunctionsPushdown extends PhysicalOptimizerRules.OptimizerRule<EvalExec> {
+
+    // TODO We could replace on filters as well
+
+    @Override
+    protected PhysicalPlan rule(EvalExec eval) {
+        return replaceSimilarityFunctionsForFieldTransformations(eval);
+    }
+
+    @SuppressWarnings("unchecked")
+    private EvalExec replaceSimilarityFunctionsForFieldTransformations(EvalExec eval) {
+        Map<Attribute, Attribute> replacements = new HashMap<>();
+        EvalExec resultEval = (EvalExec) eval.transformExpressionsDown(VectorSimilarityFunction.class, similarityFunction -> {
```

Comment:
I'd love to make this generic over other function. But I can grab that after you've landed this one. Or Alex or someone. This is a fine way to get it in, I think.